cmd/opampsupervisor/supervisor/commander/commander.go

```diff
@@ -0,0 +1,193 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commander
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/cmd/opampsupervisor/supervisor/config"
+)
+
+// Commander can start/stop/restart the Agent executable and also watch for a signal
+// for the Agent process to finish.
+type Commander struct {
+	logger  *zap.Logger
+	cfg     *config.Agent
+	args    []string
+	cmd     *exec.Cmd
+	doneCh  chan struct{}
+	waitCh  chan struct{}
+	running *atomic.Int64
+}
+
+func NewCommander(logger *zap.Logger, cfg *config.Agent, args ...string) (*Commander, error) {
+	if cfg.Executable == "" {
+		return nil, errors.New("agent.executable config option must be specified")
+	}
+
+	return &Commander{
+		logger:  logger,
+		cfg:     cfg,
+		args:    args,
+		running: &atomic.Int64{},
+	}, nil
+}
+
+// Start the Agent and begin watching the process.
+// Agent's stdout and stderr are written to a file.
+func (c *Commander) Start(ctx context.Context) error {
+	c.logger.Debug("Starting agent", zap.String("agent", c.cfg.Executable))
+
+	logFilePath := "agent.log"
+	logFile, err := os.Create(logFilePath)
+	if err != nil {
+		return fmt.Errorf("cannot create %s: %w", logFilePath, err)
+	}
+
+	c.cmd = exec.CommandContext(ctx, c.cfg.Executable, c.args...) // #nosec G204
+
+	// Capture standard output and standard error.
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21072
+	c.cmd.Stdout = logFile
+	c.cmd.Stderr = logFile
+
+	c.doneCh = make(chan struct{}, 1)
+	c.waitCh = make(chan struct{})
+
+	if err := c.cmd.Start(); err != nil {
+		return err
+	}
+
+	c.logger.Debug("Agent process started", zap.Int("pid", c.cmd.Process.Pid))
+	c.running.Store(1)
+
+	go c.watch()
+
+	return nil
+}
+
+func (c *Commander) Restart(ctx context.Context) error {
+	if err := c.Stop(ctx); err != nil {
```

Comment:
Will this work if `Stop` is called then `Restart` is called? I'm not sure what `c.cmd.Process.Signal` does if it's called multiple times in a row.

Reply:
If `Stop` is called, then `Restart` is called, the subsequent call to `Stop` is a no-op since there's a check to see whether there is a running process to stop.